[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62215
* #61942
* #61892
* __->__ #61859

Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

jerryzh168 committed:
[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar

Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -725,28 +726,40 @@ def make_input_output_share_observers(
     model: torch.nn.Module,
     modules: Dict[str, torch.nn.Module],
-) -> None:
+) -> bool:
```
nit: if we are changing this to return `bool`, can we change the name to `maybe_make_input_output_share_observers`

optional nit: some people consider it an anti-pattern to have mutations return success or failure, one could decompose this into a function which returns whether the mutation should happen (return bool) and a function which performs the mutation (return None)

jerryzh168 replied:
sure, will change the name.
For the other comment, I agree it feels a bit weird to return True/False here, but in this case it looks like the check is not trivial, do we have other ways that we can improve the structure?

vkuzo replied:
one option would be to throw an exception on failure (instead of returning False), and catch it in the caller

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -924,10 +955,11 @@ def insert_observers_for_model(
                         node_name_to_target_dtype)
 
                     is_last_node_of_pattern = root_node is node
+                    # TODO: add is_like_copy_node method in quantizehandler
```
maybe we can do this in this PR?  `qhandler.num_tensor_args` is currently not used outside of `quantization_patterns.py`, would be great to keep it private.

jerryzh168 replied:
OK, sure

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -288,10 +292,19 @@ def convert(self,
+                kwargs = load_arg(quantized=torch.float)(self.binary_op_node.kwargs)
+                op_out = quantized_graph.node_copy(self.binary_op_node, load_arg(quantized=torch.float))
+
+                def modified_load_arg(n: Node):
```
is there a way to avoid this?  Can we make the original definition of `load_arg` handle this instead?

jerryzh168 replied:
this is a hack that removes the `env` argument, I think we don't want to spend too much time on this, I think this will be gone after we do a refactor that removes `env` for convert

vkuzo replied:
makes sense, maybe we could add a comment to the original `load_arg` mentioning that some callsites may override it?

jerryzh168 replied:
OK will add this in a later PR, landing this now

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[bc-breaking] Produce quantization pattern for add_scalar and mul_scalar"


Summary:
BC-breakign note:
Previously we do not add observer/fake_quant for output of add/mul for tensor - scalar operation,
in this PR we added the observer/fake_quant instance (that's the same as input) to correctly model
the behavior of the quantized add_scalar and mul_scalar op (since quantized add/mul scalar assumes the
output quantized tensor have the same quantization parameter as input)

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29770859](https://our.internmc.facebook.com/intern/diff/D29770859)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29770859).

ejguan commented on torch/quantization/fx/prepare.py:
```diff
@@ -789,6 +806,17 @@ def make_input_output_share_observers(
+    for output_obs_node, _ in items:
+        assert is_activation_post_process_node(output_obs_node, modules)
+        output_obs_node.replace_all_uses_with(node)
+        model.graph.erase_node(output_obs_node)
```
mypy complains here.
